### PR TITLE
feat: support multiple peer selection for filter client

### PR DESCRIPTION
**Pull request by chaitanyaprem**
# Description
Addresses #1001 and provides an option for a user to specify upto how many peers a filter subscription is to be made

# Changes

- [x] Update general peer selection algo to support multiple peer selection
- [x] Update Filter client code to support subscribing and unsubscribing from multiple peers
- [x] Add tests to validate behaviour

Notes: This PR is a little messy for now and makes filter code more complex  :pray:, will refactor once we move out subscription management functionality to an SDK layer ( as prt of a new milestone)
- Peer selection with RTT still selects only single peer, this will be taken up later.
- Not supporting option to specify peers to be excluded (as we can subscribe to same peer again and the user for now can manage this) This limtation will gets addressed once we have proper subscription management implemented in filter-sdk layer.

### CODE REVIEW

File: waku/v2/peermanager/service_slot.go [hunk 1]
```diff
@@ -19,14 +19,22 @@ func newPeerMap() *peerMap {
 	}
 }
 
-func (pm *peerMap) getRandom() (peer.ID, error) {
+func (pm *peerMap) getRandom(count int) (peersMap, error) {
```
[the remark this answers is not in the document]
Comment (chaitanyaprem) [reply waku/v2/peermanager/service_slot.go:22]: Agreed, let me see if i can reuse it.

File: waku/v2/protocol/filter/client.go [hunk 2]
```diff
@@ -606,12 +629,24 @@ func (wf *WakuFilterLightNode) unsubscribeAll(ctx context.Context, opts ...Filte
...
 	peers := make(map[peer.ID]struct{})
-	subs := wf.subscriptions.GetSubscription(params.selectedPeer, protocol.ContentFilter{})
-	if len(subs) == 0 && params.selectedPeer != "" {
-		result.Add(WakuFilterPushError{
-			Err:    err,
-			PeerID: params.selectedPeer,
-		})
+	var subs []*subscription.SubscriptionDetails
+	if params.selectedPeers.Len() == 0 {
+		subs = wf.subscriptions.GetSubscription("", protocol.ContentFilter{})
```
[the remark this answers is not in the document]
Comment (chaitanyaprem) [reply waku/v2/protocol/filter/client.go:634]: It doesn't look neat, good idea. 
Will go through the code to see why this was alllowed with empty string.

File: waku/v2/protocol/filter/client.go [hunk 3]
```diff
@@ -606,12 +629,24 @@ func (wf *WakuFilterLightNode) unsubscribeAll(ctx context.Context, opts ...Filte
...
+		if len(subs) == 0 {
+			return result, nil
+		}
+	}
+	for _, peer := range params.selectedPeers {
+		subs = wf.subscriptions.GetSubscription(peer, protocol.ContentFilter{})
```
[the remark this answers is not in the document]
Comment (chaitanyaprem) [reply waku/v2/protocol/filter/client.go:640]: Good catch, will fix it.

[the remark this answers is not in the document]
Comment (chaitanyaprem) [reply waku/v2/protocol/filter/client.go]: Addressed in [26cacfe](https://github.com/waku-org/go-waku/pull/1005/commits/26cacfe4e830b05a64dd5a449077167e24c71f82)

File: waku/v2/protocol/filter/client.go [hunk 4]
```diff
@@ -606,12 +629,24 @@ func (wf *WakuFilterLightNode) unsubscribeAll(ctx context.Context, opts ...Filte
...
+		if len(subs) == 0 {
+			result.Add(WakuFilterPushError{
+				Err:    err,
```
[the remark this answers is not in the document]
Comment (chaitanyaprem) [reply waku/v2/protocol/filter/client.go]: Addressed in [26cacfe](https://github.com/waku-org/go-waku/pull/1005/commits/26cacfe4e830b05a64dd5a449077167e24c71f82)

File: waku/v2/peermanager/service_slot.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (chaitanyaprem) [reply waku/v2/peermanager/service_slot.go:22]: Addressed in [4499312](https://github.com/waku-org/go-waku/pull/1005/commits/4499312eecbefd83b2c342add63b127d79be68ef)

File: waku/v2/protocol/filter/client.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (chaitanyaprem) [reply waku/v2/protocol/filter/client.go]: Addressed little differently but in [4499312](https://github.com/waku-org/go-waku/pull/1005/commits/4499312eecbefd83b2c342add63b127d79be68ef)

File: waku/v2/protocol/legacy_filter/waku_filter.go [hunk 5]
```diff
@@ -252,7 +252,7 @@ func (wf *WakuFilter) requestSubscription(ctx context.Context, filter ContentFil
 		opt(params)
 	}
 	if wf.pm != nil && params.selectedPeer == "" {
-		params.selectedPeer, _ = wf.pm.SelectPeer(
+		selectedPeers, _ := wf.pm.SelectPeers(
```
[the remark this answers is not in the document]
Comment (chaitanyaprem) [reply waku/v2/protocol/legacy_filter/waku_filter.go:255]: This is legacy filter code while will be removed soon, hence not planning to improve it.

File: waku/v2/peermanager/peer_selection.go [hunk 6]
```diff
@@ -30,73 +32,118 @@ func (pm *PeerManager) SelectPeerByContentTopics(proto protocol.ID, contentTopic
...
+		i++
+		if i == count {
+			break
+		}
+	}
+	if len(selectedPeers) == 0 {
+		return nil, ErrNoPeersAvailable
+	}
+	return selectedPeers, nil
 }
```
Comment (richard-ramos) [inline waku/v2/peermanager/peer_selection.go:96]: No random function?

File: waku/v2/peermanager/peer_selection.go [hunk 7]
```diff
@@ -14,8 +13,11 @@ import (
 	wps "github.com/waku-org/go-waku/waku/v2/peerstore"
 	waku_proto "github.com/waku-org/go-waku/waku/v2/protocol"
 	"go.uber.org/zap"
+	"golang.org/x/exp/maps"
 )
 
+type peersMap map[peer.ID]struct{}
```
Comment (richard-ramos) [inline waku/v2/peermanager/peer_selection.go:19]: Should probably be called `peersSet`

File: waku/v2/protocol/lightpush/waku_lightpush.go [hunk 8]
```diff
@@ -264,7 +265,9 @@ func (wakuLP *WakuLightPush) handleOpts(ctx context.Context, message *wpb.WakuMe
 	}
 
 	if params.pm != nil && params.selectedPeer == "" {
-		params.selectedPeer, err = wakuLP.pm.SelectPeer(
+		var selectedPeers peer.IDSlice
+		//TODO: update this to work with multiple peer selection
+		selectedPeers, err = wakuLP.pm.SelectPeers(
```
Comment (richard-ramos) [inline waku/v2/protocol/lightpush/waku_lightpush.go:270]: `err` is not handled. Line 279 might fail then.

File: waku/v2/protocol/legacy_filter/waku_filter.go (hunk 5 above)
Comment (richard-ramos) [inline waku/v2/protocol/legacy_filter/waku_filter.go:255]: Unhanlded `err` here, might fail on line 264

File: waku/v2/protocol/filter/filter_proto_ident_test.go [hunk 9]
```diff
@@ -173,8 +173,9 @@ func (wf *WakuFilterLightNode) IncorrectSubscribe(ctx context.Context, contentFi
 	for pubSubTopic, cTopics := range pubSubTopicMap {
 		var selectedPeer peer.ID
 		//TO Optimize: find a peer with all pubSubTopics in the list if possible, if not only then look for single pubSubTopic
-		if params.pm != nil && params.selectedPeer == "" {
-			selectedPeer, err = wf.pm.SelectPeer(
+		if params.pm != nil && params.selectedPeers[0] == "" {
+			var selectedPeers peer.IDSlice
+			selectedPeers, err = wf.pm.SelectPeers(
```
Comment (richard-ramos) [inline waku/v2/protocol/filter/filter_proto_ident_test.go:178]: Unhandled `err`. Might fail on line 187

**Review by richard-ramos — APPROVED** [review]

File: waku/v2/peermanager/peer_selection.go (hunk 6 above)
Comment (chaitanyaprem) [reply waku/v2/peermanager/peer_selection.go:96]: Map's iterator already has randomness in golang and hence no need to use any additional random function.

**Pull request merged**